### PR TITLE
Add ability to toggle html/visual modes through the bridge

### PR DESCRIPTION
**Pull request by pinarol**
This change adds switch to html/visual mode messaging to the bridge and demonstrates it on the example app. 
On RN side: Updates the BlockManager to handle TOGGLE_BLOCK_MODE updates and re-render its contents.

screen shot 2018-11-26 at 16 33 52

Part of #182

**To Test**

Test 1:

- Run yarn install & yarn start and yarn ios as usual to open the example app
- Tap (...) button on the rightside of the navigation bar
- Tap "Switch to HTML"
- Verify that HTML editor is shown
- Tap (...) button on the rightsize of the navigation bar again
- Tap "Switch to Visual"
- Verify that visual editor is shown

Test 2:

- Reopen the example app
- Tap (...) button on the rightside of the navigation bar
- Tap "Switch to HTML"
- Edit something in html edit mode(do not switch back to the visual mode)
- Tap Save in the navigation bar
- Verify that you see the edited version of the html in Xcode logs

Test 3:

- Reopen the example app
- Tap (...) button on the rightside of the navigation bar
- Tap "Switch to HTML"
- Edit something in html edit mode
- Tap (...) button on the rightsize of the navigation bar again
- Tap "Switch to Visual"
- Verify that you see the edited version of the html in visual mode
- Tap Save in the navigation bar
- Verify that you see the edited version of the html in Xcode logs

Also:

- Verify that both "Inspect blocks" and "Switch to html mode" switches are removed.

### CODE REVIEW

**Comment by koke** [comment]
I pushed a commit that gets rid of the extra padding at the top, which wasn't that noticeable before, but looked weird without the toggles

File: src/components/html-text-input.js [hunk 1]
```diff
@@ -8,23 +8,50 @@ import { Platform, TextInput, KeyboardAvoidingView } from 'react-native';
...
+		};
+	}
+
+	static getDerivedStateFromProps( props: PropsType, state: StateType ) {
+		if ( state.isDirty ) {
+			return null;
+		}
+
+		return {
+			value: props.html,
```
[the remark this answers is not in the document]
Comment (pinarol) [reply src/components/html-text-input.js]: yes looks like I had forgotten to remove that

File: src/block-management/block-manager.js [hunk 2]
```diff
@@ -30,17 +30,16 @@ export type BlockListType = {
 	moveBlockDownAction: string => mixed,
 	deleteBlockAction: string => mixed,
 	createBlockAction: ( string, BlockType ) => mixed,
-	parseBlocksAction: string => mixed,
 	serializeToNativeAction: void => void,
+	toggleHtmlModeAction: void => void,
```
[the remark this answers is not in the document]
Comment (pinarol) [reply src/block-management/block-manager.js]: right, I had forgotten to remove that

File: ios/gutenberg/GutenbergViewController.swift [hunk 3]
```diff
@@ -35,3 +37,53 @@ extension GutenbergViewController: GutenbergBridgeDelegate {
...
+        let cancelAction = UIAlertAction(title: "Keep Editing", style: .cancel)
+        alert.addAction(toggleHTMLModeAction)
+        alert.addAction(cancelAction)
+        
+        present(alert, animated: true)
+    }
+    
+    func toggleHTMLMode(_ action: UIAlertAction) {
+        self.htmlMode = !self.htmlMode
+        self.gutenberg.toggleHTMLMode()
```
Comment (koke) [inline ios/gutenberg/GutenbergViewController.swift]: I don't think you need any of the `self.` here? So much JS lately I'm not even sure about Swift anymore 😱

**Comment by koke** [comment]
I tested this and works great, iOS code looks good as well.

File: ios/gutenberg/GutenbergViewController.swift (hunk 3 above)
Comment (pinarol) [reply ios/gutenberg/GutenbergViewController.swift]: no you are right, this should be a result of a copy paste from aztec side of things

File: src/block-management/block-manager.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (pinarol) [reply src/block-management/block-manager.js]: correction we use it from MainApp.js

[the remark this answers is not in the document]
Comment (Tug) [reply src/block-management/block-manager.js]: Right, let's keep it there so we can use the same propTypes here and in MainApp

File: src/components/html-text-input.js [hunk 4]
```diff
@@ -8,23 +8,39 @@ import { Platform, TextInput, KeyboardAvoidingView } from 'react-native';
...
+import { serialize, parse } from '@wordpress/blocks';
+import { withSelect, withDispatch } from '@wordpress/data';
+import { withInstanceId, compose } from '@wordpress/compose';
 
 type PropsType = {
 	blocks: Array<Object>,
-	parseBlocksAction: string => mixed,
+	onChange: string => mixed,
+	onPersist: string => mixed,
+	html: string,
```
Comment (etoledom) [inline src/components/html-text-input.js]: A question:
Do we need `html` in both props and in the component state?
Seems redundant to have it in both unless I'm missing something important 🤔

**Comment by etoledom** [comment]
A little note @hypest .
By merging this PR, we are going to lose the ability to switch to HTML in the Android example app, until  the bridge and a native-side switch is implemented there.

**Comment by etoledom** [comment]
It would be good also, at this point, to bundle a new App.js file, since the current one is quite outdated and WPiOS won't recognize the new bridge message without this code.

But it could be on a follow up PR too. 👍
cc @koke

File: src/components/html-text-input.js (hunk 4 above)
Comment (pinarol) [reply src/components/html-text-input.js]: no you are right, removing it from props

**Comment by koke** [comment]
I'll be pushing a new version of the bundle tomorrow with all the latest changes

**Pull request merged**